src/common/hist_util.cc

```diff
@@ -186,6 +188,107 @@ class GHistBuildingManager {
+  svfloat64_t grad = svdup_n_f64(p_gpair[idx_gh]);
+  svfloat64_t hess = svdup_n_f64(p_gpair[idx_gh + 1]);
+
+  for (size_t j = 0; j < row_size; j += svcntw()) {
+    // Create a predicate (mask) for 32-bit & 64-bit elements, active only for valid elements
+    svbool_t pg32 = svwhilelt_b32(j, row_size);
+    svbool_t pg64_lower = svwhilelt_b64(j, row_size);
+    svbool_t pg64_upper = svwhilelt_b64(j+svcntd(), row_size);
+
+    // Load the gradient index values and offsets for the current chunk of the row
+    svuint32_t gr_index_vec = load_index_vec(pg32, &gr_index_local[j]);
+    svuint32_t idx_bin_vec;
+    if (kAnyMissing) {
+      idx_bin_vec = svmul_n_u32_x(pg32, gr_index_vec, two);
+    } else {
+      svuint32_t offsets_vec = svld1(pg32, &offsets[j]);
+      svuint32_t temp = svadd_u32_m(pg32, gr_index_vec, offsets_vec);
+      idx_bin_vec = svmul_n_u32_x(pg32, temp, two);
+    }
+
+    // Unpack 32-bit index binary vector into 64-bit vectors from lower & upper half respectively
+    svuint64_t idx_bin_vec0_0 = svunpklo_u64(idx_bin_vec);
+    svuint64_t idx_bin_vec0_1 = svunpkhi_u64(idx_bin_vec);
+
+    // Increment the indices by 1 for hessian.
+    svuint64_t idx_bin_vec1_0 = svadd_n_u64_m(pg64_lower, idx_bin_vec0_0, 1);
+    svuint64_t idx_bin_vec1_1 = svadd_n_u64_m(pg64_upper, idx_bin_vec0_1, 1);
+
+    // Gather the histogram data corresponding to the computed indices
+    svfloat64_t hist0_vec0 = svld1_gather_index(pg64_lower, hist_data, idx_bin_vec0_0);
+    svfloat64_t hist0_vec1 = svld1_gather_index(pg64_upper, hist_data, idx_bin_vec0_1);
+    svfloat64_t hist1_vec0 = svld1_gather_index(pg64_lower, hist_data, idx_bin_vec1_0);
+    svfloat64_t hist1_vec1 = svld1_gather_index(pg64_upper, hist_data, idx_bin_vec1_1);
+
+    // Accumulate the gradient and hessian values into the histogram
+    hist0_vec0 = svadd_f64_m(pg64_lower, hist0_vec0, grad);
+    hist0_vec1 = svadd_f64_m(pg64_upper, hist0_vec1, grad);
+    hist1_vec0 = svadd_f64_m(pg64_lower, hist1_vec0, hess);
+    hist1_vec1 = svadd_f64_m(pg64_upper, hist1_vec1, hess);
+
+    // Store the updated histogram data back into memory
+    svst1_scatter_index(pg64_lower, hist_data, idx_bin_vec0_0, hist0_vec0);
+    svst1_scatter_index(pg64_upper, hist_data, idx_bin_vec0_1, hist0_vec1);
+    svst1_scatter_index(pg64_lower, hist_data, idx_bin_vec1_0, hist1_vec0);
+    svst1_scatter_index(pg64_upper, hist_data, idx_bin_vec1_1, hist1_vec1);
+  }
+}
+#endif
+
+// Returns true if SVE ISA is available on the current CPU (with caching)
+#ifdef __linux__
+int check_sve_hw_support() {
+    static int cached_sve_support = -1;
+    if (cached_sve_support == -1) {
+        int ret = prctl(PR_SVE_GET_VL);
+        if (ret == -1) {
+            cached_sve_support = 0;
+        } else {
+            cached_sve_support = 1;
+        }
+    }
+    return cached_sve_support;
+}
+
+static int sve_enabled = check_sve_hw_support();
```

Comment:
Does the value of a global static variable valid when accessed from multiple threads? It might be better to thread-local storage instead.

@trivialfis Any thoughts on this topic?

Reply:
I will work on it. Still learning the code.